// @iden3/js-iden3-auth/dist/cjs/circuits/atomicMtp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtomicQueryMTPPubSignals = void 0;
const core_1 = require("../core/core");
const id_1 = require("../core/id");
const query_1 = require("./query");
const ownershipVerifier_1 = require("./ownershipVerifier");
class AtomicQueryMTPPubSignals extends ownershipVerifier_1.IDOwnershipPubSignals {
    constructor(pubSignals) {
        super();
        if (pubSignals.length != 74) {
            throw new Error(`invalid number of Output values expected ${74} got ${pubSignals.length}`);
        }
        const userIdBytes = core_1.Core.intToBytes(BigInt(pubSignals[0]));
        this.userId = id_1.Id.idFromBytes(userIdBytes);
        this.userState = BigInt(pubSignals[1]);
        this.challenge = BigInt(pubSignals[2]);
        this.issuerClaimIdenState = BigInt(pubSignals[3]);
        const issuerIdBytes = core_1.Core.intToBytes(BigInt(pubSignals[4]));
        this.issuerId = id_1.Id.idFromBytes(issuerIdBytes);
        this.issuerClaimNonRevState = BigInt(pubSignals[5]);
        this.timestamp = parseInt(pubSignals[6], 10);
        this.claimSchema = BigInt(pubSignals[7]);
        this.slotIndex = parseInt(pubSignals[8], 10);
        this.operator = parseInt(pubSignals[9], 10);
        this.values = [];
        for (let index = 0; index < 64; index++) {
            const val = pubSignals[10 + index];
            this.values.push(BigInt(val));
        }
    }
    async verifyQuery(query, schemaLoader) {
        const outs = {
            issuerId: this.issuerId.string(),
            schemaHash: this.claimSchema,
            operator: this.operator,
            slotIndex: this.slotIndex,
            value: this.values,
        };
        return await (0, query_1.checkQueryRequest)(query, outs, schemaLoader);
    }
    async verifyStates(resolver) {
        const userStateResolved = await resolver.resolve(this.userId.bigInt(), this.userState);
        if (!userStateResolved.latest) {
            throw new Error(`only latest states are supported`);
        }
        const issuerStateResolved = await resolver.resolve(this.issuerId.bigInt(), this.issuerClaimIdenState);
        if (!issuerStateResolved) {
            throw new Error(`issuer state is not valid`);
        }
        const issuerNonRevStateResolved = await resolver.resolve(this.issuerId.bigInt(), this.issuerClaimNonRevState);
        if (!issuerNonRevStateResolved.latest &&
            Date.now() -
                issuerNonRevStateResolved.transitionTimestamp * 1000 >
                60 * 60 * 1000) {
            throw new Error(`issuer state for non-revocation proofs is not valid`);
        }
        return;
    }
}
exports.AtomicQueryMTPPubSignals = AtomicQueryMTPPubSignals;
