// @iden3/js-iden3-auth/dist/cjs/circuits/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthPubSignals = void 0;
const core_1 = require("../core/core");
const id_1 = require("../core/id");
const ownershipVerifier_1 = require("./ownershipVerifier");
class AuthPubSignals extends ownershipVerifier_1.IDOwnershipPubSignals {
    constructor(pubSignals) {
        super();
        if (pubSignals.length != 3) {
            throw new Error(`invalid number of Output values expected ${3} got ${pubSignals.length}`);
        }
        this.challenge = BigInt(pubSignals[0]);
        this.userState = BigInt(pubSignals[1]);
        const bytes = core_1.Core.intToBytes(BigInt(pubSignals[2]));
        this.userId = id_1.Id.idFromBytes(bytes);
    }
    async verifyQuery(_query) {
        throw new Error('Method not implemented.');
    }
    async verifyStates(resolver) {
        const userStateResolved = await resolver.resolve(this.userId.bigInt(), this.userState);
        if (!userStateResolved.latest) {
            throw new Error(`only latest states are supported`);
        }
        return;
    }
}
exports.AuthPubSignals = AuthPubSignals;
