// @iden3/js-iden3-auth/dist/cjs/circuits/query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkQueryRequest = void 0;
const keccak256_1 = __importDefault(require("keccak256"));
const nested_property_1 = __importDefault(require("nested-property"));
const util_1 = require("../core/util");
const operators = new Map([
    ['$noop', 0],
    ['$eq', 1],
    ['$lt', 2],
    ['$gt', 3],
    ['$in', 4],
    ['$nin', 5],
]);
const serializationIndexDataSlotAType = 'serialization:IndexDataSlotA';
const serializationIndexDataSlotBType = 'serialization:IndexDataSlotB';
const serializationValueDataSlotAType = 'serialization:ValueDataSlotA';
const serializationValueDataSlotBType = 'serialization:ValueDataSlotB';
async function checkQueryRequest(query, outputs, schemaLoader) {
    const issuerAllowed = query.allowedIssuers.some((issuer) => issuer === '*' || issuer === outputs.issuerId);
    if (!issuerAllowed) {
        throw new Error('issuer of claim is not allowed');
    }
    const loadResult = await schemaLoader.load(query.schema);
    if (loadResult.extension !== 'json-ld') {
        throw new Error('only json-ld schema is supported');
    }
    const toHash = new Uint8Array([
        ...loadResult.schema,
        ...toBytes(query.schema.type),
    ]);
    const schemaHash = (0, keccak256_1.default)(Buffer.from(toHash));
    // only json ld-schema are supported
    const cq = parseRequest(query.req, loadResult.schema, query.schema.type, outputs.value.length);
    if (outputs.operator !== cq.operator) {
        throw new Error(`operator that was used is not equal to request`);
    }
    if (outputs.operator === operators.get('$noop')) {
        // for noop operator slot and value are not used in this case
        return;
    }
    if (outputs.slotIndex !== cq.slotIndex) {
        throw new Error(`wrong claim slot was used in claim`);
    }
    if (outputs.operator !== cq.operator) {
        throw new Error(`operator that was used is not equal to request`);
    }
    for (let index = 0; index < cq.values.length; index++) {
        if (outputs.value[index].toString(10) !== cq.values[index].toString(10)) {
            throw new Error(`comparison value that was used is not equal to requested in query`);
        }
    }
    const shBigInt = (0, util_1.fromLittleEndian)(schemaHash.slice(Math.ceil(schemaHash.length / 2), schemaHash.length));
    if (shBigInt.toString() !== outputs.schemaHash.toString()) {
        throw new Error(`schema that was used is not equal to requested in query`);
    }
    return;
}
exports.checkQueryRequest = checkQueryRequest;
const toBytes = (s) => {
    //TODO: buffer is not present in browser
    const buffer = Buffer.from(s, 'utf8');
    const result = Uint8Array.from(buffer);
    return result;
};
function parseRequest(req, schema, credType, valueLength) {
    if (!req) {
        return {
            operator: operators.get('$noop'),
            values: null,
            slotIndex: 0,
        };
    }
    let fieldName = '';
    let fieldReq;
    if (Object.keys(req).length > 1) {
        throw new Error(`multiple requests  not supported`);
    }
    for (const [key, value] of Object.entries(req)) {
        fieldName = key;
        fieldReq = value;
        if (Object.keys(fieldReq).length > 1) {
            throw new Error(`multiple predicates for one field not supported`);
        }
        break;
    }
    const slotIndex = getFieldSlotIndex(fieldName, credType, schema);
    let operator;
    const values = new Array(valueLength).fill(BigInt(0));
    for (const [key, value] of Object.entries(fieldReq)) {
        if (!operators.has(key)) {
            throw new Error(`operator is not supported by lib`);
        }
        operator = operators.get(key);
        if (Array.isArray(value)) {
            for (let index = 0; index < value.length; index++) {
                values[index] = BigInt(value[index]);
            }
        }
        else {
            values[0] = BigInt(value);
        }
        break;
    }
    const cq = {
        operator,
        values,
        slotIndex,
    };
    return cq;
}
function getFieldSlotIndex(fieldName, credentialType, schema) {
    const obj = JSON.parse(Buffer.from(schema).toString('utf-8'));
    const type = nested_property_1.default.get(obj, `@context.0.${credentialType}.@context.${fieldName}.@type`);
    switch (type) {
        case serializationIndexDataSlotAType:
            return 2;
        case serializationIndexDataSlotBType:
            return 3;
        case serializationValueDataSlotAType:
            return 6;
        case serializationValueDataSlotBType:
            return 7;
        default:
            return -1;
    }
}
