// @iden3/js-iden3-auth/dist/cjs/state/resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isGenesisStateId = exports.EthStateResolver = void 0;
const core_1 = require("../core/core");
const util_1 = require("../core/util");
const ethers_1 = require("ethers");
const abi_1 = require("./abi");
class EthStateResolver {
    constructor(rpcUrl, contractAddress) {
        this.rpcUrl = rpcUrl;
        this.contractAddress = contractAddress;
    }
    async resolve(id, state) {
        let url = new URL(this.rpcUrl);
        const ethersProvider = new ethers_1.ethers.providers.JsonRpcProvider({
            url: url.href,
            user: url.username,
            password: url.password,
        });
        const contract = new ethers_1.ethers.Contract(this.contractAddress, abi_1.stateABI, ethersProvider);
        const isGenesis = isGenesisStateId(id, state);
        if (isGenesis) {
            return { latest: true, state, transitionTimestamp: 0 };
        }
        const contractState = await contract.getState(id);
        if (contractState.toBigInt() === 0n) {
            throw new Error('state is not found. Identity is not genesis');
        }
        if (contractState.toBigInt() !== state) {
            const transitionInfo = await contract.getTransitionInfo(state);
            if (transitionInfo[4].toBigInt() === 0n) {
                throw new Error('Transition info contains invalid id');
            }
            if (transitionInfo[0].toBigInt() === 0n) {
                throw new Error('No information of transition for non-latest state');
            }
            return {
                latest: false,
                state: state,
                transitionTimestamp: transitionInfo[0].toBigInt(),
            };
        }
        return { latest: true, state, transitionTimestamp: 0 };
    }
}
exports.EthStateResolver = EthStateResolver;
function isGenesisStateId(id, state) {
    const idBytes = (0, util_1.toLittleEndian)(id, 31);
    const typeBJP0 = new Uint8Array(2);
    const stateBytes = (0, util_1.toLittleEndian)(state, 32);
    const idGenesisBytes = stateBytes.slice(-27);
    // we take last 27 bytes, because of swapped endianness
    const idFromStateBytes = Uint8Array.from([
        ...typeBJP0,
        ...idGenesisBytes,
        ...core_1.Core.calculateChecksum(typeBJP0, idGenesisBytes),
    ]);
    if (JSON.stringify(idBytes) !== JSON.stringify(idFromStateBytes)) {
        return false;
    }
    return true;
}
exports.isGenesisStateId = isGenesisStateId;
